README.md
### NCICharts: Linear charts and charts with range selectors for iOS

1. [Instalation](#instalation)
1. [Simple](#simple)
1. [Zoom smooth line and axis customization](#zoom-smooth-line-and-axis-customization)
1. [Ranges](#ranges)
1. [Customization options](#customization-options)
1. [Ranges chart customization](#ranges-chart-customization)
1. [Live updates](#live-updates)

## Instalation

```
pod 'NCICharts'
```

## Simple

![alt text][Illustration]
[Illustration]: https://raw.github.com/FlowForwarding/dynamiccharts/master/docs/simple.png "NCI simple chart"

```ObjectiveC
#import "NCISimpleChartView.h"

    NCISimpleChartView *chart = [[NCISimpleChartView alloc] initWithFrame:CGRectMake(50, 30, 400, 250)];
    [self.view addSubview:chart];
    
    int numOfPoints = 10;
    for (int ind = 0; ind < numOfPoints; ind ++){
        if (ind % 5 == 0){
            [chart addPoint:ind val:@[[NSNull null]]];
        } else {
            [chart addPoint:ind val:@[@(arc4random() % 5)]];
        }
    }
``` 

## Zoom smooth line and axis customization

![alt text][IllustrationZoom]
[IllustrationZoom]: https://raw.github.com/FlowForwarding/dynamiccharts/master/docs/zoom.gif 

![alt text][IllustrationDecr]
[IllustrationDecr]: https://raw.github.com/FlowForwarding/dynamiccharts/master/docs/decr.png 

```ObjectiveC
    NCISimpleChartView *zoomingChart =  [[NCISimpleChartView alloc] initWithFrame:
                                         CGRectMake(self.view.frame.size.height + 50, 30, 400, 250)
                                                                       andOptions:@{nciGraphRenderer: [NCIZoomGraphView class],
                                                                                    nciIsSmooth: @[@YES],
                                                                                    nciYAxis: @{
                                                                                            nciAxisShift: @320,
                                                                                            nciInvertedLabes: @YES,
                                                                                            nciLineDashes: @[],
                                                                                            nciLineWidth: @2
                                                                                            },
                                                                                    nciXAxis: @{
                                                                                            nciLineColor: [UIColor greenColor],
                                                                                            nciLineWidth: @2,
                                                                                            nciAxisShift : @90,
                                                                                            nciInvertedLabes: @YES
                                                                                            }}];
    
    [pages addSubview:zoomingChart];
    
    int numOfPoints = 10;
    for (int ind = 0; ind < numOfPoints; ind ++){
        [zoomingChart addPoint:ind val:@[@(arc4random() % 5)]];
    }
```

```ObjectiveC
    NCISimpleChartView *nciChartDecreasing =  [[NCISimpleChartView alloc] initWithFrame:
                                               CGRectMake(self.view.frame.size.height + 50, 30, 400, 250) andOptions:
                                               @{
                                                 nciGraphRenderer: [NCIZoomGraphView class],
                                                 nciLineWidths: @[@2],
                                                 nciIsFill: @[@(NO)],
                                                 nciYAxis: @{
                                                         nciAxisDecreasing: @YES,
                                                         nciAxisShift : @320,
                                                         nciInvertedLabes: @YES
                                                         },
                                                 nciXAxis: @{
                                                         nciAxisDecreasing: @YES,
                                                         nciAxisShift : @0,
                                                         nciInvertedLabes: @YES
                                                         }
                                                 }];
    
    nciChartDecreasing.topBottomGridSpace = 0;
    
    [pages addSubview:nciChartDecreasing];
    
    for (int ind = 10; ind < 50; ind ++){
        [nciChartDecreasing addPoint:ind val:@[@(ind*ind)]];
    }
```

## Ranges

![alt text][Illustration2]
[Illustration2]: https://raw.github.com/FlowForwarding/dynamiccharts/master/docs/ranges.png "NCI chart"

```ObjectiveC
#import "NCIChartView.h"

    NCIChartView *chart = [[NCIChartView alloc] initWithFrame:CGRectMake(0, 0, 300, 400)];
    [self.view addSubview:chart];
    int numOfPoints = 10;
    for (int ind = 0; ind < numOfPoints; ind ++){
        [chart addPoint:ind val:@[@(arc4random() % 5)]];
    }
    chart.minRangeVal = 5;
    chart.maxRangeVal = 8; 
```

## Customization options

![alt text][Illustration4]
[Illustration4]: https://raw.github.com/FlowForwarding/dynamiccharts/master/docs/cuatomization.png "NCI customized chart" 

```ObjectiveC
    NCISimpleChartView *simpleChart = [[NCISimpleChartView alloc]
                   initWithFrame:CGRectMake(50, 30, 400, 250)
                   andOptions: @{
                                 
                                 nciIsFill: @[@(NO), @(NO)],
                                 nciIsSmooth: @[@(NO), @(YES)],
                                 nciLineColors: @[[UIColor orangeColor], [NSNull null]],
                                 nciLineWidths: @[@2, [NSNull null]],
                                 nciHasSelection: @YES,
                                 nciSelPointColors: @[[UIColor redColor]],
                                 nciSelPointImages: @[[NSNull null], @"star"],
                                 nciSelPointTextRenderer: ^(double argument, NSArray* values){
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"yyyy-MMM-dd HH:mm:ss"];
        return [[NSAttributedString alloc] initWithString:
                [NSString stringWithFormat:@"y: %.0f,%.0f  x:%@",
                 [values[0] floatValue],
                 [values[1] floatValue],
                 [dateFormatter stringFromDate:[NSDate dateWithTimeIntervalSince1970: argument]]]
                                               attributes: @{NSForegroundColorAttributeName: [UIColor redColor],
                                                             NSFontAttributeName: [UIFont fontWithName:@"MarkerFelt-Thin" size:12]}];
    },
                                 
                                 nciSelPointSizes: @[@10, [NSNull null]],
                                 
                                 //                                               nciTapGridAction: ^(double argument, double value, float xInGrid, float yInGrid){
                                 //
                                 //    },
                                 nciShowPoints : @YES,
                                 nciUseDateFormatter: @YES,//nciXLabelRenderer
                                 nciXAxis: @{nciLineColor: [UIColor redColor],
                                             nciLineDashes: @[],
                                             nciLineWidth: @2,
                                             nciLabelsFont: [UIFont fontWithName:@"MarkerFelt-Thin" size:12],
                                             nciLabelsColor: [UIColor blueColor],
                                             nciLabelsDistance: @120,
                                             nciUseDateFormatter: @YES
                                             },
                                 nciYAxis: @{nciLineColor: [UIColor blackColor],
                                             nciLineDashes: @[@2,@2],
                                             nciLineWidth: @1,
                                             nciLabelsFont: [UIFont fontWithName:@"MarkerFelt-Thin" size:12],
                                             nciLabelsColor: [UIColor brownColor],
                                             nciLabelsDistance: @30,
                                             nciLabelRenderer: ^(double value){
        
        return [[NSAttributedString alloc] initWithString: [NSString stringWithFormat:@"%.1f$", value]
                                               attributes:@{NSForegroundColorAttributeName: [UIColor brownColor],
                                                            NSFontAttributeName: [UIFont fontWithName:@"MarkerFelt-Thin" size:12]}];
    },
                                             },
                                 nciGridVertical: @{nciLineColor: [UIColor purpleColor],
                                                    nciLineDashes: @[],
                                                    nciLineWidth: @1},
                                 nciGridHorizontal: @{nciLineColor: [UIColor greenColor],
                                                      nciLineDashes: @[@2,@2],
                                                      nciLineWidth: @2},
                                 nciGridColor: [[UIColor magentaColor] colorWithAlphaComponent:0.1],
                                 nciGridLeftMargin: @50,
                                 nciGridTopMargin: @50,
                                 nciGridBottomMargin: @40
                                 }];
    
    simpleChart.backgroundColor = [[UIColor yellowColor] colorWithAlphaComponent:0.2];
    
    int numOfPoints = 10;
    double dataPeriod = 60*60*24*30;
    double step = dataPeriod/(numOfPoints - 1);
    for (int ind = 0; ind < numOfPoints; ind ++){
        //to use default date formatter for Y axis, provide arguments as  timeIntervalSince1970
        //and set nciXLabelRenderer option to YES
        [simpleChart addPoint:[[NSDate dateWithTimeIntervalSinceNow: -dataPeriod + step *ind] timeIntervalSince1970] val:@[@(arc4random() % 5),
                                  @(arc4random() % 5)]];
    }            

``` 

## Ranges chart customization

```ObjectiveC

    NCIChartView *chart = [[NCIChartView alloc]
                           initWithFrame:CGRectMake(0, 0, 400, 250)
                           andOptions: @{
                                         nciTopGraphOptions : @{
                                                 //simple chart options
                                                 },
                                         nciBottomGraphOptions : @{
                                                 //simple chart options
                                                 },
                                         nciLeftRangeImageName: @"image",
                                         nciRightRangeImageName: @"image",
                                         nciBottomChartHeight: @50
                                         }
                           ];
//@property(nonatomic)double minRangeVal;
//@property(nonatomic)double maxRangeVal;
//@property (nonatomic, copy) void (^rangesMoved)(void);   

```

## Live updates

```ObjectiveC

    NCISimpleChartView *chart = [[NCISimpleChartView alloc]
                                 initWithFrame:CGRectMake(50, 30, 400, 250)];
    [self.view addSubview:chart];
    int numOfPoints = 10;
    for (int ind = 0; ind < numOfPoints; ind ++){
        [chart addPoint:ind val:@[@(arc4random() % 5)]];
    }
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        for (int ind = numOfPoints; ind < 4*numOfPoints; ind ++){
            [NSThread sleepForTimeInterval:2.0f];
            [chart addPoint:ind val:@[@(2*ind)]];
            dispatch_async(dispatch_get_main_queue(), ^{
                [chart drawChart];
            });
        }
    });
```    



CHANGELOG.md
= 1.0.0 (2014-01-29)

 * Core for dynamic charts with number of customization options

= 1.0.1 (2014-01-30)

 * Range selector chart right range setter correction

= 1.0.2 (2014-03-26)

 * Update nciIsFill property form bool to array of bools to control if graph line is filled for each series
 
 * Fixed issue with calculating Y axis max and min values for multiple lines zoom chart

 * Fixed shifting from time to time selected points on chart during fast scrolling 
 
= 1.0.3 (2014-04-01)
 
 * Fix for shifted X-labels on simple chart
 
 * New sapmle

= 1.0.4 (2014-04-01)

 * Option to use NSAttributedString for selected point label
 
 * Fix issue for hiding/showing selection label
 
 * Fix for shifted X-labels
 
= 1.0.5 (2014-04-09) 

 * API to set selected point programmatically

 * Working test project

 * Fix for shifted X-labels

= 2.0.0

 * API improvements

 * Smooth(bezier) line support

 * More axis cutomizataions(move axis and their labels)

= 2.0.1

 *Fix x-axis labels
 
 *Add customization option for decreasing axis


NCIChart/views/NCIAxis.m
//
//  NCIAxis.m
//  NCIChart
//
//  Created by Ira on 4/15/14.
//  Copyright (c) 2014 FlowForwarding.Org. All rights reserved.
//

#import "NCIAxis.h"
#import "NCIChartOptions.h"
#import "NCISimpleGraphView.h"
#import "NCISimpleGridView.h"


//TODO make 2 objects for NCIXAxis and NCIYAxis
@interface NCIAxis()

@property(nonatomic)bool invertedLabes;
@property(nonatomic, strong)UIFont* labelsFont;
@property(nonatomic, strong)UIColor* labelsColor;
@property(nonatomic, copy) NSString* (^labelRenderer)(double);
@property(nonatomic)bool nciUseDateFormatter;
@property(nonatomic)float labelsDistance;
@property(nonatomic)float nciAxisShift;

@property(nonatomic)float labelWidth;
@property(nonatomic)float labelHeight;

@property(nonatomic, strong) NSDateFormatter *dateFormatter;
@property(nonatomic) double dimention;
@property(nonatomic) double step;

@end

@implementation NCIAxis

- (id)initWithOptions:(NSDictionary *)options{
    if (self = [super initWithOptions:options]){
        _nciAxisShift = NAN;
        _dateFormatter = [[NSDateFormatter alloc] init];
        _labels = [[NSMutableArray alloc] init];
        
        if ([options objectForKey:nciInvertedLabes])
            _invertedLabes = [[options objectForKey:nciInvertedLabes] boolValue];
        
        if ([options objectForKey:nciUseDateFormatter])
            _nciUseDateFormatter = [[options objectForKey:nciUseDateFormatter] boolValue];
        
        if ([options objectForKey:nciLabelRenderer])
            _labelRenderer = [options objectForKey:nciLabelRenderer] ;
        
        if ([options objectForKey:nciAxisDecreasing])
            _nciAxisDecreasing = [[options objectForKey:nciAxisDecreasing] boolValue];
        
        if ([options objectForKey:nciLabelsFont]){
            _labelsFont = [options objectForKey:nciLabelsFont];
        } else {
             if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad){
               _labelsFont = [UIFont italicSystemFontOfSize:14];
             } else {
                _labelsFont = [UIFont italicSystemFontOfSize:10];
             }
        }
        
        if ([options objectForKey:nciLabelsColor]){
            _labelsColor = [options objectForKey:nciLabelsColor];
        } else {
            _labelsColor = [UIColor blackColor];
        }
        
        
        if ([options objectForKey:nciAxisShift])
            _nciAxisShift = [[options objectForKey:nciAxisShift] floatValue];
        
        if ([options objectForKey:nciLabelsDistance]){
            _labelsDistance = [[options objectForKey:nciLabelsDistance] floatValue];
        } else {
            if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad){
                 _labelsDistance = 150;
            } else {
                 _labelsDistance = 60;
            }
        }
    }
    return self;
}

- (void)setChart:(NCISimpleChartView *)chart{
    _chart = chart;
    if (self.vertical){
        _labelHeight = _labelsDistance;
        if (_nciAxisShift == _nciAxisShift){
            _labelWidth = _nciAxisShift + self.chart.nciGridLeftMargin;
        } else {
            _labelWidth = self.chart.nciGridLeftMargin;
        }
    } else {
        _labelHeight =  self.chart.nciGridBottomMargin;
        _labelWidth =  _labelsDistance;
    }
}

- (void)redrawLabels:(float)length min:(double)min max:(double)max{
    
    _dimention = length;
    for (UILabel *label in _labels){
        [label removeFromSuperview];
    }
     [_labels removeAllObjects];
     _step = length/(max - min);
    [self formatDateForDistance];
    
    if (self.vertical){
        if (!self.chart.nciHasHorizontalGrid)
            return;
        for(int i = 0; i<= length/_labelsDistance; i++){
            float yPos = length - i*_labelsDistance - _labelsDistance/2 + self.chart.nciGridTopMargin;
            double curVal = [self.chart.graph getValByY: _labelsDistance*i];
            UILabel *label;
            if (_invertedLabes){
                label = [[UILabel alloc] initWithFrame:
                                  CGRectMake(_nciAxisShift + self.chart.nciGridLeftMargin + 4,
                                                                            yPos, _labelWidth, _labelHeight)];
                label.textAlignment = NSTextAlignmentLeft;
            } else {
                label = [[UILabel alloc] initWithFrame: CGRectMake(0, yPos, _labelWidth - 4, _labelHeight)];
                label.textAlignment = NSTextAlignmentRight;
            }
            [self makeUpLabel:label val:curVal];
        }
    } else {
        for(int i = 0; i< length/_labelsDistance; i++){
            float xPos = self.chart.nciGridLeftMargin + _labelsDistance *i;
            if (xPos > (length + self.chart.nciGridLeftMargin - _labelsDistance/2))
                continue;
            float yPos = (_nciAxisShift != _nciAxisShift) ?
                self.chart.graph.frame.size.height + self.chart.nciGridTopMargin :
                _nciAxisShift + self.chart.nciGridTopMargin;
            UILabel *label;
            if (_invertedLabes){
                label = [[UILabel alloc] initWithFrame:
                         CGRectMake(xPos, yPos - _labelHeight, _labelWidth, _labelHeight)];
            } else {
                label = [[UILabel alloc] initWithFrame:
                         CGRectMake(xPos, yPos, _labelWidth, _labelHeight)];
            }
            double curVal = [self.chart.graph getArgumentByX: (_labelsDistance *i + _labelWidth/2)];
            label.textAlignment = NSTextAlignmentCenter;
            [self makeUpLabel:label val:curVal];
        }
    }
    
}

- (void)makeUpLabel:(UILabel *)label val:(double) curVal{
    label.textColor = _labelsColor;
    label.font = _labelsFont;
    if (_labelRenderer){
        NSObject *value =  _labelRenderer(curVal);
        if ([[value class] isSubclassOfClass:[NSAttributedString class]]){
            label.attributedText = (NSAttributedString *)value;
        } else {
            label.text = (NSString *)value;
        }
    } else if (self.nciUseDateFormatter){
        label.text = [_dateFormatter stringFromDate: [NSDate dateWithTimeIntervalSince1970:curVal]];
    } else {
        label.text = [NSString stringWithFormat:@"%0.1f", curVal];
    }
    [self.labels addObject:label];
    [self.chart addSubview:label];
    
}

- (void)formatDateForDistance{
    if ((1/_step * _labelsDistance) < 60*60*24){
        [_dateFormatter setDateFormat:@"yyyy-MMM-dd HH:mm"];
    } else if ((1/_step * _labelsDistance) < 60*60*24*30){
        [_dateFormatter setDateFormat:@"yyyy-MMM-dd"];
    } else {
        [_dateFormatter setDateFormat:@"yyyy-MMM"];
    }
}

- (void)drawBoundary:(CGContextRef ) currentContext{
    [self setUpLine:currentContext];
    if (_vertical){
        if (_nciAxisShift != _nciAxisShift){
             _nciAxisShift = 0;
        }
        CGContextMoveToPoint(currentContext, _nciAxisShift, 0);
        CGContextAddLineToPoint(currentContext, _nciAxisShift, _dimention);
        CGContextStrokePath(currentContext);
    } else {
        if (_nciAxisShift != _nciAxisShift)
            _nciAxisShift = self.chart.graph.grid.frame.size.height;
        CGContextMoveToPoint(currentContext, 0, _nciAxisShift);
        CGContextAddLineToPoint(currentContext, _dimention, _nciAxisShift);
        CGContextStrokePath(currentContext);
    }
}

@end
